### PR TITLE
Fast(er) Generation of Sonar Pointclouds 

### PR DESCRIPTION
**Pull request by marc-micatka**
## Overview
Updating callback method in sonar_pointcloud.py to numpy vectorization for that sweet speed.
Method is also compatible with 8bit/32bit data input and log scales the output:
## Results
**32bit Real Data**
image

**8bit from Gazebo Simulation**
image

### CODE REVIEW

File: ros/include/sonar_image_proc/sonar_image_msg_interface.h [hunk 1]
```diff
@@ -56,7 +71,11 @@ struct SonarImageMsgInterface
     } else if (_ping->data_size == 2) {
       return intensity_uint16(idx) >> 8;
     } else if (_ping->data_size == 4) {
-      return intensity_uint32(idx) >> 24;
+      if (do_log_scale_) {
```
Comment (lauralindzey) [inline ros/include/sonar_image_proc/sonar_image_msg_interface.h:74]: Looks like log scale only works on 32 bit data and not the others??

File: ros/include/sonar_image_proc/sonar_image_msg_interface.h [hunk 2]
```diff
@@ -84,6 +107,17 @@ struct SonarImageMsgInterface
...
+        return intensity_float(idx) * UINT32_MAX;
+      } else {
+        return intensity_uint32_nonlog(i);
+      }
+    }
+    return 0;
+  }
+
+  uint32_t intensity_uint32_nonlog(const size_t i) const {
```
Comment (lauralindzey) [inline ros/include/sonar_image_proc/sonar_image_msg_interface.h:119]: OK, I see what this is used for, but the whole calculation-of-logs things seems like a mess. 

There's a flag controlling whether intensity_uint32 calculates logs, but there also a separate function called intensity_uint32_nonlog. 

And logs are apparently only calculated for 32-bit values.

Is there maybe a single place where *all* scaling could be done, perhaps after getting the intensity into a float value for plotting?

File: scripts/sonar_pointcloud.py [hunk 3]
```diff
@@ -11,53 +11,81 @@
...
         self.elev_step = np.radians(elev_step_deg)
-        self.elevations = np.arange(self.min_elev, self.max_elev+self.elev_step, self.elev_step)
+        self.elevations = np.arange(self.min_elev,
+                                    self.max_elev + self.elev_step,
+                                    self.elev_step)
 
         # TODO: assert that this is in range of uint8? (Or, msg.data_size)
-        self.intensity_threshold = rospy.get_param("~intensity_threshold", 100)
+        self.intensity_threshold = rospy.get_param("~intensity_threshold",
+                                                   0.74)
```
Comment (lauralindzey) [inline scripts/sonar_pointcloud.py:42]: If this has jumped from 8 bit counts to [0, 1.0], update the comment.

File: scripts/sonar_pointcloud.py [hunk 4]
```diff
@@ -11,53 +11,81 @@
...
+                                    self.callback,
+                                    queue_size=1,
+                                    buff_size=1000000)
         self.pub = rospy.Publisher("sonar_cloud", PointCloud2, queue_size=1)
 
-        min_elev_deg = rospy.get_param("~min_elev_deg", 0)
-        max_elev_deg = rospy.get_param("~max_elev_deg", 0)
-        assert(max_elev_deg >= min_elev_deg)
-        elev_step_deg = rospy.get_param("~elev_step_deg", 5.0)
+        min_elev_deg = rospy.get_param("~min_elev_deg", -10)
```
Comment (lauralindzey) [inline scripts/sonar_pointcloud.py:29]: @marc-micatka -- do you know how to specify rosparams on the command line? 
(Also, do you know about `git add -p`?)

File: scripts/sonar_pointcloud.py [hunk 5]
```diff
@@ -73,60 +101,94 @@ def make_geometry(self, image_msg):
                     zz = distance * se
                     points[kk][idx] = [xx, yy, zz]
 
-        self.geometry = points
+        self.geometry = np.array(points)
+        # Pre-allocate our output points
```
Comment (lauralindzey) [inline scripts/sonar_pointcloud.py:105]: How does this work with your Zulip comments about image size changing mid-dataset?

**Comment by lauralindzey** [comment]
@marc-micatka -- I took a quick look. I think the biggest problem I have with this code was in the part you didn't write =)

I'm happy to merge this whenever. I don't care as much about code quality for one-off utilities as I do for code that is core to our research ... for something like this, the important bit is that it works, as you demoed earlier today =)

File: ros/include/sonar_image_proc/sonar_image_msg_interface.h (hunk 2 above)
Comment (micat001) [reply ros/include/sonar_image_proc/sonar_image_msg_interface.h:119]: Yeah that would be better for sure. I intentionally just changed the script which was taking in a plain sonar message, similar to how it was originally written.

File: scripts/sonar_pointcloud.py (hunk 4 above)
Comment (micat001) [reply scripts/sonar_pointcloud.py:29]: I do, is it better to set the defaults via the launch file instead of here?

**Comment by micat001** [comment]
closing in favor of [20](https://github.com/apl-ocean-engineering/sonar_image_proc/pull/20)

Based on feature branch, rebranching from main.,

File: scripts/sonar_pointcloud.py (hunk 5 above)
Comment (micat001) [reply scripts/sonar_pointcloud.py:105]: I check the shape on each callback and re-compute make_geometry() if the shape has changed

**Pull request closed without merging**